### ISSUE TITLE
Multi-value attributes

### ISSUE THREAD

**Issue by staplegun** [ckeditor/ckeditor5#1339]
## Is this a bug report or feature request? (choose one)

🆕 Feature request 

## 💻 Version of CKEditor

CKEditor 5 v11.0.0

## 📃 Other details that might be useful

Hi, I want to attach multiple instances of the same attribute to a piece of text, each with a different embedded value.

I think this may be a feature request as I don't think this is currently supported, but possibly there is another way of achieving the outcome I am after.

What I am trying to achieve:

1. I want to add semantic markup inside the HTML/model that indicates the topic various sentences are about, something like this:
```HTML
<p>
  <subject sid="activity-1">
    <subject sid="relation-1">
      Tom saw Mary
    </subject>
    eating 
    <subject sid="food-1">
      pepperoni pizza
    </subject>
  </subject>
</p>
```

2. However, I want the markup to be 'sticky', e.g. if I cut the word "pizza" and paste it somewhere else in the document, I want 'pizza' to still have subject id 'food-1' attached in its new location.

From my reading of the documentation, attributes stay with each word as they are moved around, but container elements (like 'subject') are not. 

For me, even though attributes are intended for presentation, they seem to fit my requirements better, e.g. 
```
Tom saw Mary = subject(activity-1), subject(relation-1)
eating = subject(activity-1)
pepperoni pizza = subject(activity-1), subject(food-1)
```

However, it looks like attributes are singular and only contain a true/false embedded value, whereas I need each attribute to also have a value embedded (the subject id) and for a text piece to have multiple subject attributes.

I'd be happy to be told that there already support for multiple attributes/values that isn't documented, and/or there is another part of the CKE framework that would meet my needs.

**Comment by Reinmar** [comment]
This is a very interesting case.

First of all, multi-value attributes could be applied in this way:

```js
writer.createText( 'Tom saw Mary', { subject: [ 'activity-1', 'relation-1' ] } );
```

That creates a text node with an attribute which value is an array. The tricky part is to convert it correctly to the view to create a nested structure like the one you proposed.

To do that you'd need to write a custom converter for [`DowncastDispatcher#attribute`](https://ckeditor.com/docs/ckeditor5/latest/api/module_engine_conversion_downcastdispatcher-DowncastDispatcher.html#event-attribute). Since this is just one attribute, this will be a one converter (for `#attribute:subject`). However, it needs to apply as many `AttributeElement` as the number of items in the array. You apply attribute elements by calling [`DowncastWriter#wrap()`](https://ckeditor.com/docs/ckeditor5/latest/api/module_engine_view_downcastwriter-DowncastWriter.html#function-wrap) in your converter.

`AttributeElement` has a [`#priority`](https://ckeditor.com/docs/ckeditor5/latest/api/module_engine_view_attributeelement-AttributeElement.html#member-priority) which defines how attribute elements are applied in the view. Those with lower priority number will wrap those with higher priority number (e.g. `<a>` has `5` and `<strong>` has `10` so they'll be applied like this: `<a><strong>foo</strong></a>`). Also, attribute elements with the same priority and name will be merged.

So, since you want `<subject sid="relation-1">` to be inside `<subject sid="activity-1">`, it should have a higher priority number. 

Additionally, if you need more control over merging check out [`AttributeElement#isSimilar()`](https://ckeditor.com/docs/ckeditor5/latest/api/module_engine_view_attributeelement-AttributeElement.html#function-isSimilar)'s docs.

So, that was about downcasting. Without implementing it I can't tell if it's really doable, but it certainly seems so.

But that's just half of the story. Now you also need to handle upcasting (loading data, pasting). So you need upcast converters which can do the opposite process – based on multiple elements, create one model attribute. Again, this is rather doable, but requires a custom converter and some thinking.

tl;dr This should be doable but without a deeper knowledge about various conversion mechanisms, it's hard to understand how the process should look.

**Comment by scofalik** [comment]
> From my reading of the documentation, attributes stay with each word as they are moved around, but container elements (like 'subject') are not.

I think you missed a point here a little.

In the model, text nodes may have attributes. For example (using your notation):

```
Tom saw Mary = bold(true)
```

And then they are converted to the view (and then, to the DOM):

```
<strong>Tom saw Mary<strong>
```

So if your `<subject>` feature is implemented as attributes it will move around when the letters are moved. So you are fine in this regard.

As you correctly deduced, you would need to set multi-value attributes on those texts and handle conversion. Attributes aren't really only for presentation purposes, so you are right to use them as you wanted. For example, we use attributes to set links on text.

@Reinmar already did a good job describing how to handle downcast. For upcasting, you can base on the code that already exists in `upcast-converters.js` in `_prepareToAttributeConverter`. Now, understanding what happens there is complicated, to be honest, but the only difference would be in `_setAttributeOn` function where instead of setting the attribute, you would have to see if `section` attribute is already set and if it is, add to the array, instead of re-setting the attribute:

```
const sectionValue = data.viewItem.getAttribute( 'sid' );
const currentValue = node.getAttribute( 'section' );
let newValue;

if ( currentValue ) {
    // I am not sure if this is needed but I copy `currentValue` on purpose
    // to be sure that the conversion will be properly fired.
    // To be tested if that's needed at all. Maybe you could just go with
    // `currentValue.push( sectionValue )` and be done.
    newValue = currentValue.slice();
    newValue.push( sectionValue );
} else {
    newValue = [ sectionValue ];
}

conversionApi.writer.setAttribute( 'section', sectionValue, node );
```

`_prepareToAttributeConverter` is a generalized converter which is quite complicated because it has to be configurable through config passed in plain object. Your converter would be simpler than that. The steps that you need to do for sure is:

1. Test if consumable was consumed. Return if it was, consume if it wasn't.
2. Get `sid` value from `data.viewItem`.
3. Check if `data.modelRange` is available, if not, fire children conversion as in the original function.
4. Use `conversionApi.writer` to set model attribute on `data.modelRange`.

Not sure if I am missing something.

Here I wrote a lot about the theory behind how attributes work: https://stackoverflow.com/questions/51303892/how-to-add-target-attribute-to-a-tag-in-ckeditor5/51365516#51365516

I don't have enough time to write it for you, but together with insights from @Reinmar you should be able to do it alone. 🤞

**Comment by staplegun** [comment]
Thanks @Reinmar and @scofalik, your tips got me going in the right direction. The upcast conversion was easy, except I then got lost trying to create two elements in the view during downcast, and I'm not sure if it's something I'm doing wrong or the current codebase doesn't support it?

1. Upcast

```javascript
// subject/src/subjectcommand.js
// based on @ckeditor/ckeditor5-basic-styles/src/attributecommand.js
export default class SubjectCommand extends Command {
	constructor( editor, attributeKey ) {
		...
	}

	refresh() {
		...
	}

	execute( options = {} ) {
		const model = this.editor.model;
		const doc = model.document;
		const selection = doc.selection;

		const subjectId = prompt( 'Subject ID' );
		const commandValue = subjectId;

		// add new value to array if not already present
		// if is present, remove it (assumes repeat means disable)
		let newValue;
		if ( this.value ) {
			newValue = this.value.slice();
			let position = newValue.indexOf( commandValue );
			if (position === -1) {
				newValue.push( commandValue );
			} else {
				newValue.splice( position, 1 );
			}
		} else {
			newValue = [ commandValue ];
		}

		model.change( writer => {
			if ( selection.isCollapsed ) {
				if ( newValue ) {
					writer.setSelectionAttribute( this.attributeKey, newValue );
				} else {
					writer.removeSelectionAttribute( this.attributeKey );
				}
			} else {
				const ranges = model.schema.getValidRanges( selection.getRanges(), this.attributeKey );
				for ( const range of ranges ) {
					if ( newValue ) {
						writer.setAttribute( this.attributeKey, newValue, range );
					} else {
						writer.removeAttribute( this.attributeKey, range );
					}
				}
			}
		} );
	}

	_getValueFromFirstAllowedNode() {
		// NB: this is tweaked from 
		//   return item.hasAttribute( this.attributeKey );
		// to
		//   return item.getAttribute( this.attributeKey );
	}
}

```

2. Downcast option 1

I tried adding two converters to the model change event to create two of the same (`subject`) view elements.

When this didn't work, I dropped back to just trying to create *any* two completely different view elements (`alpha` and `beta`) elements. I assume this should work as `downcastMarkerToHighlight()` does this. But even this didn't work - it only ever creates one view element (the one with the higher priority).

3. Downcast option 2

So I then tried a single converter that creates the two view elements within it.  Again this didn't work - it only ever creates the first view element in the code.

Basically, I can't work out how to create two view elements from a single model event...

```javascript
// subject/src/subjectediting.js
// based on @ckeditor/ckeditor5-image/src/imageediting.js

import Plugin from '@ckeditor/ckeditor5-core/src/plugin';
import { downcastAttributeToElement } from '@ckeditor/ckeditor5-engine/src/conversion/downcast-converters';
import { wrap } from '@ckeditor/ckeditor5-engine/src/conversion/downcast-converters';
import { upcastAttributeToAttribute } from '@ckeditor/ckeditor5-engine/src/conversion/upcast-converters';
import { cloneDeep } from 'lodash-es';
// TODO downcast-converters.js tries to import as ModelSelection, which doesn't exist
import Selection from '@ckeditor/ckeditor5-engine/src/model/selection';
import DocumentSelection from '@ckeditor/ckeditor5-engine/src/model/documentselection';
import SubjectCommand from './subjectcommand';

const SUBJECT_COMMAND = 'subject';
const SUBJECT_MODEL_ATTRIBUTE = 'subject';
const SUBJECT_VIEW_ELEMENT = 'subject';
//TODO it looks like view attributes can't be specified via a variable
//const SUBJECT_VIEW_ID = 'sid';

export default class SubjectEditing extends Plugin {
	 init() {
		const editor = this.editor;

		// Allow subject attribute on text nodes.
		editor.model.schema.extend( '$text', { allowAttributes: SUBJECT_MODEL_ATTRIBUTE } );

		// converter from model to view
		editor.conversion.for( 'downcast' ).add( this.downcastAttributeToMultipleElements( {
			model: SUBJECT_MODEL_ATTRIBUTE,
			view: {
				name: SUBJECT_VIEW_ELEMENT,
				key: 'sid'
			}
		}));

		// converter from view to model
		editor.conversion.for( 'upcast' ).add( upcastAttributeToAttribute( {
				view: {
					name: SUBJECT_VIEW_ELEMENT,
					key: 'sid'
				},
				model: SUBJECT_MODEL_ATTRIBUTE,
				converterPriority: 'low'
		} ) );

		// Create subject command.
		editor.commands.add( SUBJECT_COMMAND, new SubjectCommand( editor, SUBJECT_COMMAND ) );
	}

	downcastAttributeToMultipleElements( config ) {
		config = cloneDeep( config );
		let eventName = 'attribute:' + SUBJECT_MODEL_ATTRIBUTE;

		// TODO loop through model attribute values to create creators, for debugging using hard-coded elements
		let elementCreator1 = ( modelData, viewWriter ) => makeElementCreator( config, viewWriter, 'id123', 10, 'alpha' );
		let elementCreator2 = ( modelData, viewWriter ) => makeElementCreator( config, viewWriter, 'id456', 20, 'beta' );

		return dispatcher => {

			// uncomment which option is being tested
			
			// option 1 - attach two converters to event (each create single element)
			// dispatcher.on( eventName, wrap( elementCreator1 ), { 'priority': 10 } );
			// dispatcher.on( eventName, wrap( elementCreator2 ), { 'priority': 20 } );
			
			// option 2 - single converter that creates two elements
			// dispatcher.on( eventName, makeElements( config ), { 'priority': 10 } );
		};
	}
}

export function makeElementCreator(config, viewWriter, value, priority, name ) {
	return viewWriter.createAttributeElement( name, {'sid': value}, {'priority': priority} );
}

export function makeElements(config) {
	return ( evt, data, conversionApi ) => {
		let elementCreator1 = ( modelData, viewWriter ) => makeElementCreator( config, viewWriter, 'id123', 10, 'alpha' );
		let elementCreator2 = ( modelData, viewWriter ) => makeElementCreator( config, viewWriter, 'id456', 20, 'beta' );
		wrapInner( evt, data, conversionApi, elementCreator1 );
		wrapInner( evt, data, conversionApi, elementCreator2 );
	}
}

// copied from @ckeditor/ckeditor5-engine/src/conversion/downcast-converters.js
// NB: removed wrapper function parameters leaving just the inner function
export function wrapInner( evt, data, conversionApi, elementCreator ) {
	...
}
```

**Comment by scofalik** [comment]
Adding two downcast converters surely won't work. They will simply override and the one with bigger priority will "win". It is like this so it is possible to extend/overwrite already existing features.

What kind of structure exactly you try to build? It will be easier for me to help you if you will show me what you want to try to build rather than try to deduce it from the code.

**Comment by staplegun** [comment]
Thanks @scofalik.  I'm trying to insert semantic elements, which in HTML would all be the same element with different attribute values, but may also be embedded inside each other. For example:

```HTML
<p>
  <subject sid="activity-1">
    <subject sid="relation-1">
      Tom saw Mary
    </subject>
    eating 
    <subject sid="food-1">
      pepperoni pizza
    </subject>
  </subject>
</p>
```

In effect, this is fairly similar to span elements, for example when encoding RDFa:

```HTML
<p>
  My 
  <span property="foaf:interest">
    favorite book is the inspiring 
    <span about="urn:ISBN:0752820907">
      <span property="dc:title">Weaving the Web</span> by
      <span property="dc:creator">Tim Berners-Lee</span>
    </span>
  </span>.
</p>
```

I started off encoding these in the CKEditor model as attributes on $text, as this is more semantically correct - 'creator' and 'of interest' are properties of Tim Berners-Lee, rather than saying there 15 characters (Tim B-L) that currently happen to be inside a wider range of characters that have a start and end for presentation purposes.

I also liked the idea that if the user cut&pasted "Tim Berners-Lee", then both the creator and interest properties would automatically travel with that text.

So I modeled the CKE model for a piece of text as containing `attribute:subject[ 'activity-1', 'relation-1' ]`. Then the downcasting would create two wrapping `<subject sid='...'>` elements (I was hoping the auto-merge would merge adjacent subject elements together). 

I can see elsewhere in the CKE codebase that `span` elements are created, but only one at a time - I believe I would need to convert the array of model attribute values into separate view elements in a single downcast conversion.

I'm assuming that if instead I modeled the subjects as wrapping element ranges, I would then need to dig into the code that executes during cut&paste and look around to find all the ranges that this pieve of text is within? I was hoping to code within an existing pattern instead of having to find all the related code for various possible scenarios.

**Comment by scofalik** [comment]
Okay, so the structure is same as in the original post. I thought it changed based on your previous post.

In this case, you should not need to do any trickery with multiple converters. Instead, you need a single converter that will set all the attributes/value in your subject array.

For this, I'd advise writing your own converter instead of using conversion helpers.

```javascript
editor.conversion.for( 'downcast' ).add( dispatcher => {
    dispatcher.on( 'attribute:' + SUBJECT_MODEL_ATTRIBUTE, ( evt, data, conversionApi ) => {
        // Step 1 - check consumables (check if the attribute wasn't handled already).
        if ( conversionApi.consumable.consume( data.item, evt.name ) ) {
            return;
        }

        // Step 2 - get view range on which attribute(s) should be changed.
        const viewRange = conversionApi.mapper.toViewRange( data.range );

        // Step 3 - manage attributes in view. This has to be done in two steps.
        // First - remove old attributes. Then - add new attributes.
        // For example, old subject value could be `[ 'foo' ]` while new is `[ 'bar' ]`.
        // So you not only need to add the new value but also remove the old.
        for ( const value of data.attributeOldValue ) {
            const viewElement = viewWriter.createAttributeElement( 'subject', { 'sid': value } );
            conversionApi.writer.unwrap( viewElement, viewRange );
        }

        for ( const value of data.attributeNewValue ) {
            const viewElement = viewWriter.createAttributeElement( 'subject', { 'sid': value } );
            conversionApi.writer.wrap( viewElement, viewRange );
        }

        evt.stop();
    } );
} );
```

I am writing this from the top of my head so it may contain bugs but this is a starting point. Maybe this will be enough. Let me know.

**Comment by scofalik** [comment]
If you want to learn more about API you can check [conversion.DowncastDispatcher](https://ckeditor.com/docs/ckeditor5/latest/api/module_engine_conversion_downcastdispatcher-DowncastDispatcher.html) (and all the conversion APIs, honestly) and also [view.DowncastWriter](https://ckeditor.com/docs/ckeditor5/latest/api/module_engine_view_downcastwriter-DowncastWriter.html) (it is a view writer used during downcast conversion, available at `conversionApi.writer`).

**Comment by Kushan-** [comment]
hey speaking of createAttributeElement, how can we add inline style to it?? like < "<span style="color: #000000;"> >
Furthermore, after creating the attribute element how to append it to the writer?? as in 
< writer.append( tagName, docFrag ) >
I am sorry if these two are naive question but i am unable to find the answer. Any help will be much appreciated.

**Comment by Mgsy** [comment]
Closing it due to a lack of activity.
